src/neo4j_genai/llm/base.py

```diff
@@ -13,11 +13,21 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 from abc import ABC, abstractmethod
+from typing import Optional, Any
 
 
-class LLM(ABC):
+class LLMInterface(ABC):
     """Interface for large language models."""
 
+    def __init__(
+        self,
+        model_name: str,
+        model_params: Optional[dict[str, Any]] = None,
+        **kwargs: Any,
+    ):
+        self.model_name = model_name
+        self.model_params = model_params or {}
+
     @abstractmethod
     def invoke(self, input: str) -> str:
         """Sends a text input to the LLM and retrieves a response.
```

Comment:
I actually think we want to change this interface to follow langchain ChatModel instead of LLM. See https://python.langchain.com/v0.2/docs/integrations/chat/ vs https://python.langchain.com/v0.2/docs/integrations/llms/ to be able to support the latest models from OpenAI for example.
Chat models returns a `BaseMessage` (with a `content` field) instead of a `string`.
https://github.com/langchain-ai/langchain/blob/8e89178047a68f4b6519e646cfe1c96d6e58b7cc/libs/core/langchain_core/language_models/chat_models.py#L166 vs https://github.com/langchain-ai/langchain/blob/8e89178047a68f4b6519e646cfe1c96d6e58b7cc/libs/core/langchain_core/language_models/llms.py#L1252

I had to create a wrapper to use GPT-4o with text2cypher here: https://github.com/oskarhane/graphrag-org-chart/blob/fcbf6838bb70e2ca8b321b84104a6c96a2848216/src/main.py#L25

The "Completion API" at OpenAI is marked as "Legacy", and that's what being used when importing `OpenAI` rather than `ChatOpenAI` from `langchain_openai`.
https://platform.openai.com/docs/guides/text-generation/completions-api. Also see warning on https://python.langchain.com/v0.2/docs/integrations/llms/openai/.

Sorry for the wall of text, but I went down this rabbit hole yesterday when I wanted to use the newer models.
Ping to @alexthomas93 as well, as a FYI.

Reply:
Does that mean that the only supported implementations will be from langchain? I mean, without having to implement the interface by yourself?

Reply:
It doesn't mean that we _only_ support implementations from langchain, but I think we have a lot to gain by supporting their interface.
We can implement our own LLM providers we want to have first class, like OpenAI.

Reply:
Done in https://github.com/neo4j/neo4j-genai-python/pull/50/commits/6c83290798443e6e93d3f417457106af91924021